### PR TITLE
Library: Public library running arbitrary Linux 64-bit process in isolation with resource limits.

### PR DESCRIPTION
**Pull request by P-A-R-U-S**
PR contains implements for library that running arbitrary Linux 64-bit process in isolation with resource limits. For details, please, follow this RFD Document: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/blob/library/docs/RFD-JobWorkerService.md

To run test you can use following make command:

* `make run_test_cgroup` - run sunset of tests for cgroups
* `make run_test_output` - run subset of test for OutputReader and all related to Steam command output
* `make run_test_job` - run sunset of tests for Job executing (you can check Test Plan section in RFD Document)

_**Note**: because `make run_test_job` need to create and manages crgoup, you possibly need [sudo](https://en.wikipedia.org/wiki/Sudo) permission to run tests._

Package structure and files name are self-explained, but here is some details: 

```
 |-Makefile (
 |-pkg
 | |-jobWorker
 | | |-namespaces
 | | | |-cgroup.go               --> constains all cgroup functions
 | | | |-cgroup_test.go
 | | | |-mount.go
 | | |-job.go              --> contains all Start/Stop/Stream/Status implementation
 | | |-job_test.go
 | | |-outputReadCloser.go      --> implements io.ReadCloser interface to read from the provided CommandOutput
 | | |-outputReadCloser_test.go
 | | |-commandOutput.go               --> implements io.Closer and io.Writer and implement a command buffer
 | | |-commandOutput_test.go
```

### CODE REVIEW

File: pkg/jobWorker/outputReadCloser.go
Comment (greedy52) [inline pkg/jobWorker/outputReadCloser.go]: This doesn't look right to me. Closing the reader shouldn't close the buffer. 

Also, ideally closing the reader should stop any running `Read`. How do you plan to handle the case where a client stops streaming (while the process is still running)?

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: This will result in the process using the library terminating.
Why not simply return an error and allow the caller to decide how to handle it?

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: This could lead to a potential race condition when using the library:

```
status := job.Status()
if status.State == JOB_STATUS_RUNNING { 
}
```

Since status.State refers to the same memory address, it could be modified by other functions (e.g., job.status.State = JOB_STATUS_TERMINATED) without proper locking, leading to unpredictable behavior during job handling.

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: This error hanging is quite confusing Does internal server error should be propagate to the user  and mixed with job error ?

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]:
```suggestion
func (job *Job) Stream() io.ReadCloser {
```

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: Due to job.mutex.Lock()  call  In this case the job operation will be locked for 10 seconds.

File: pkg/jobWorker/job.go
Comment (greedy52) [inline pkg/jobWorker/job.go]: why `os.Exit` here?

File: pkg/jobWorker/job.go
Comment (greedy52) [inline pkg/jobWorker/job.go]: leftover cgroups on these failure cases.

File: pkg/jobWorker/job.go
Comment (greedy52) [inline pkg/jobWorker/job.go]: while waiting, `job.Status()` is blocked.

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go:344]: What if the process has children. Does the job.cmd.Process.Signal will also stop/terminate all process children's  ?

File: pkg/jobWorker/namespaces/cgroup.go
Comment (greedy52) [inline pkg/jobWorker/namespaces/cgroup.go]: why `FILE_MODE_EVERYONE` instead of `FILE_MODE_WEB`?

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky agree. Make sense.

Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @greedy52 Removed. Here is same comment about that: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2#discussion_r1744135003

File: pkg/jobWorker/namespaces/cgroup.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/namespaces/cgroup.go]: @greedy52 changes to FILE_MODE_WEB. 

`Explanation:` I've got 'permission denied' error during my cgroup testing and tried to create it with 'everyone read and write' permission, but I've changes to  FILE_MODE_WEB, retest and look ok for me.

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @greedy52 good point. I made appropriate changes.
Screenshot 2024-09-04 at 11 03 45 AM


here is testing
Screenshot 2024-09-04 at 11 01 27 AM

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @greedy52 ok. I think I can change Stop method like one the picture below and wait what would happened first SIGTEM or command completion.  

**Explanation**: I don't want to design Stop function as fire-and-forget and wrap all waiting logic in separate go-routine. It definitely would help to prevent blocking all other operations, but Stop method not able to return any error if, by any reason, process would be KILLed .  

Screenshot 2024-09-04 at 11 43 17 AM

Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: What do you think ?

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky Changed. Please, check this comment: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2#discussion_r1744272688

File: pkg/jobWorker/outputReadCloser.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/outputReadCloser.go]: > Also, ideally closing the reader should stop any running Read. How do you plan to handle the case where a client stops streaming (while the process is still running)?

@greedy52  we've discussed this with @smallinsky  before. Please, check this comment:  https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/1#discussion_r1732132687

Comment (P-A-R-U-S) [reply pkg/jobWorker/outputReadCloser.go]: > This doesn't look right to me. Closing the reader shouldn't close the buffer.

@greedy52  Yes, make sense. Added Flag to monitor reader state.

Screenshot 2024-09-04 at 1 49 59 PM

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go:344]: @smallinsky  That really interesting question. After research I've found we just need to send negative PID to KILL all child process. See here: https://linux.die.net/man/2/kill

```
If pid equals -1, then sig is sent to every process for which the 
calling process has permission to send signals, 
except for process 1 (init), but see below.

If pid is less than -1, then sig is sent to every process in 
the process group whose ID is -pid. 
```
And changes would looks like 
Screenshot 2024-09-04 at 2 23 17 PM

But such things should be deeply tested. What do you think ?

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky good point. The best option in this case create set of internal flags and based on them return appropriate status.

Screenshot 2024-09-04 at 1 49 59 PM

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky Ok. changed in more appropriate error handling and propagate error to the library consumer.
Screenshot 2024-09-04 at 6 16 14 PM

File: pkg/jobWorker/job.go
Comment (smallinsky) [reply pkg/jobWorker/job.go:344]: Please use your personal judgment if you tested this manually and it works for purpose of this challangage test coverage is not needed. 

However, I believe you need to set:
```
&syscall.SysProcAttr{
  Setpgid: true
}
 ```
 
to make it work.

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: This seems to be a potential race condition. The entire function `func (job *Job) Start() error {` is called under the lock, but the go func() { can be scheduled after the function exit, and the lock is released by the defer `job.mutex.Unlock()` call.

File: pkg/jobWorker/outputReadCloser.go
Comment (smallinsky) [inline pkg/jobWorker/outputReadCloser.go]: In this design the Read and Close calls are not thread safe. 

So let say that consumer want to create a read loop that will block by r.Read call this a new output is available but  want to cancel the reeder loop 
```go 
ctx, cancel := ctx.WithCancel(ctx)
defer cancel()

r : NewOutputReadCloser()

go func() {
   switch {
    case <- ctx.Done:
        r.Close()
   } 
}
for  { 

  .., ... = r.Read(buff)
  ... 
}

```

So the question is how a consumer of outputReadCloser can cancel ongoing stream in the Read/Close  if `outputReadCloser `  Read Close functions are not thread safe ?

File: pkg/jobWorker/outputReadCloser.go
Comment (smallinsky) [inline pkg/jobWorker/outputReadCloser.go]: Nit: https://go.dev/wiki/CodeReviewComments#receiver-names

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: nit: https://google.github.io/styleguide/go/decisions#constant-names

**Review by smallinsky — COMMENTED** [review]
I left some additional comments, but it feels like you’re almost there.

File: pkg/jobWorker/job.go
Comment (greedy52) [inline pkg/jobWorker/job.go]:
```suggestion
		cleanCGroup <- true
```

File: pkg/jobWorker/job.go
Comment (greedy52) [reply pkg/jobWorker/job.go]: I think it depends on the purpose of the lock. If the lock is meant to prevent other API calls, then it's doing its job. If the lock is meant to protect internal members like status, it's not doing anything while waiting so might as well just unlock during that. I will leave it to you to judge that.

**Review by greedy52 — APPROVED** [review]
LGTM once Marek's comment is resolved

File: pkg/jobWorker/job.go
Comment (greedy52) [inline pkg/jobWorker/job.go:329]: nit: what happens if `Stop` is called after the cmd is terminated

Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go:329]: @greedy52 returns ErrJobAlreadyStopped.
Screenshot 2024-09-05 at 11 34 39 AM

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky That from Java world I think - in Oracle they forced us to names constant like _(should be all uppercase with words separated by underscores ("_"))_: https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

Changed to team preferred format :))

Screenshot 2024-09-05 at 11 45 07 AM

File: pkg/jobWorker/outputReadCloser.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/outputReadCloser.go]: Renamed.

File: pkg/jobWorker/job.go
Comment (greedy52) [reply pkg/jobWorker/job.go:329]: do you need lock for `job.isTerminated` and `job.isCompleted`?

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky sorry, trying to understand your point about `can be scheduled after the function exit` 

I think I can create some implementation where waiting `go func()` starting after `Start` function, but in any case `go func()` waiting when 'Start' release 'Lock' to continue. See on the picture below.
Screenshot 2024-09-05 at 1 17 58 PM

And I run intentionally `processState, err := job.cmd.Process.Wait()` before 'job.mutex.Lock()' to prevent lock on other functions and lef them execute unlocked  
 
```go
		processState, err := job.cmd.Process.Wait()
		job.processState = processState

		job.mutex.Lock()
		defer job.mutex.Unlock()
```

Is that ok ?

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go:329]: Sorry, incorrect screenshot - it was moved after mutex.Lock
Screenshot 2024-09-05 at 3 27 37 PM

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @greedy52  in general I would agree with you and probably for production solution better to design `Stop` as fire-and-forget and let caller decide - would he like to wait for Termination/Completions or not.

But I'm worrying I would be able to complete challenge on time. If you don't mind - I would add TODO: and if I have time - I would try to review it completing client/server.
 
```go
	err := testJob.Stop()
	for status := testJob.Status(); status.State != JobStatusTerminated && status.State != JobStatusCompleted) {
			fmt.Println(status.State)
			time.Sleep(1 * time.Second)
	}
```
Here is  TODO.

Screenshot 2024-09-05 at 4 04 32 PM

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go:344]: @smallinsky yes, changed. It is much elegance solutions than mine.  
Thank you.

File: pkg/jobWorker/outputReadCloser.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/outputReadCloser.go]: @smallinsky  Looks like RWMutex are best solutions in this case. 
Screenshot 2024-09-05 at 6 32 07 PM

And just to prove it -  I've created one more unit test ([Test_OutputReadCloser_Expecting_calling_Close_from_different_goroutine_close_Reader](https://github.com/P-A-R-U-S/Go-Job-Worker-Service/blob/7d875c186cbd50bfc516d255adb26edd04ddf2ee/pkg/jobWorker/outputReadCloser_test.go#L156)):

1. Create Cancel context
2. Run go-routine with infinite loop to write into output 
3. Run go-routine to receive Cancel status 
4. Run go-routine to trigger Cancel after 100ms sleep
5. Run infinite loop in main function to read output.
6. Expected: Read should stop reading data after triggering Cancel and calling Close function.

As you can see on screenshot below after receiving `cancel` both reader stop processing data.
Screenshot 2024-09-05 at 6 45 52 PM

File: pkg/jobWorker/job.go
Comment (smallinsky) [reply pkg/jobWorker/job.go]: >sorry, trying to understand your point about can be scheduled after the function exit

Sorry for confusion.

The question is: what lock protects the following `job.exitReason` write [operation](https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2/files/51b5bbd76770e793822084936875536f669a2fb3#diff-35dec03abce473a482ac214d8104767d523b60d5cdf7d1248f7e02e24a8ccd6aR194)?

```go
if err := ns.DeleteCGroup(job.getCGroupName()); err != nil {
	log.Printf("error closing cgroup: %s\n", err)
	job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing cgroup: %w\n", err))
}
```

The current implementation assumes that the lock acquired by:
```go
  processState, err := job.cmd.Process.Wait()
  job.mutex.Lock()
  defer job.mutex.Unlock()
  ```
protects the job.exitReason write, but the lock can be released earlier, and the following:

```
job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing cgroup: %w\n", err))
```
could be written after the lock is released.

Imagine following scenarion where the goroutine
```
  defer func() {
    cleanCGroup <- true
    unmount <- true
  }()
  ```

is delayed by GO  scheduler and in meantime a the job.Status() is called where exitReason is read under read lock `job.getExitReason(),`  

```go
func Start() {
  job.mutex.Lock()
  defer job.mutex.Unlock()
 ...
    go func() {
        select {
        case <-unmount:
            if err := ns.UnmountProc(); err != nil {
                log.Printf("error unmounting /proc - %s\n", err)
                job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error unmounting /proc - %w\n", err))
            }
        }
    }()
 ...
    go func() {
        processState, err := job.cmd.Process.Wait()
        job.mutex.Lock()
        defer job.mutex.Unlock()
        job.isCompleted = true

        defer func() {
            sleep(time.Second * 10)
            cleanCGroup <- true
            unmount <- true
        }()
        
        if err = job.output.Close(); err != nil {
            job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing output: %w\n", err))
        }
        return
    }()
 ...
 return
}

```

The current flow doesn't quarantine the that `job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing output: %w\n", err))` is called under the lock

Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: @smallinsky Thank you Marek for details - now I got your point. As I see there are two possible options to resolved it.

Using `biderectional channel` and add `defer` functions to wait cleanup status. Because `defer` execute in LIFO order - all of them would be executed before we actually release Lock.

Screenshot 2024-09-06 at 1 15 34 PM

And it is quite tricky to design unit test to check this, but I've `artificially` slow down go-routines which `deleting cgroups/unmount`, added log and here is executions process as a proof. 

Screenshot 2024-09-06 at 1 33 31 PM


2: Other possible options using watGroup, but looks like we end-up with case when `Start` functions blocked and waiting until `cgroups/mount` clean-up ether from`Start` function or from `go func()`. And some other refactoring probably need. As for me it is looks complicated and potentially introduce extra effort to support it in future.  

Let me know what do you think ?

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: This is dirty approach where the time Sleep is used to artificially delay operation to force  particular condition for UT.

File: pkg/jobWorker/job.go
Comment (smallinsky) [inline pkg/jobWorker/job.go]: Race condition those writes are unprotected by any lock  and it the a user calls job.Status will cause data race.

File: pkg/jobWorker/job.go
Comment (smallinsky) [reply pkg/jobWorker/job.go]: > And it is quite tricky to design unit test to check this, but I've artificially slow down go-routines which

https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2/files#diff-35dec03abce473a482ac214d8104767d523b60d5cdf7d1248f7e02e24a8ccd6aR203

This seem to be very hack way to test things caused by overcomplicated code. 

If you need to call 
```go
 if err {
    cleanCGroup <- true
     waitingCleanCGroupToCompleted 
	  defer waitingCleanCGroupToCompleted()
		return fmt.Errorf("error creating cgroup: %w", err)
}
```
why not just  move cgrup into function and do the cleanup synchronous  ?

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: Sorry, it was done for testing purpose to prove this: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2#discussion_r1747710054 and I've removed `sleep`.

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: Resolved: Moved after Lock to prevent race conditions.

File: pkg/jobWorker/job.go
Comment (P-A-R-U-S) [reply pkg/jobWorker/job.go]: Done. Agree code looks much easy.

**Review by smallinsky — APPROVED** [review]

**Review by greedy52 — APPROVED** [review]

**Pull request merged**
